elixir/apps/domain/lib/domain/cache/client.ex

```diff
@@ -0,0 +1,557 @@
+          r
+        else
+          r.id not in added_ids
+        end
+      end)
+
+    cache = %{cache | connectable_resources: connectable_resources}
+
+    {:ok, added, Enum.map(removed, & &1.id) |> Enum.map(&load!/1), cache}
+  end
+
+  @doc """
+    Fetches a membership id by an actor_group_id.
+  """
+
+  @spec fetch_membership_id(t(), Cache.Cacheable.uuid_binary()) ::
+          {:ok, Ecto.UUID.t()} | {:error, :not_found}
+
+  def fetch_membership_id(cache, gid_bytes) do
+    cache.memberships
+    |> Map.fetch(gid_bytes)
+    |> case do
+      {:ok, mid_bytes} -> {:ok, load!(mid_bytes)}
+      :error -> {:error, :not_found}
+    end
+  end
+
+  @doc """
+    Adds a new membership to the cache, potentially fetching the missing policies and resources
+    that we don't already have in our cache.
+
+    Since this affects connectable resources, we recompute the connectable resources, which could
+    yield deleted IDs, so we send those back.
+  """
+
+  @spec add_membership(t(), Clients.Client.t()) ::
+          {:ok, [Domain.Cache.Cacheable.Resource.t()], [Ecto.UUID.t()], t()}
+
+  def add_membership(cache, client) do
+    # TODO: Optimization
+    # For simplicity, we rehydrate the cache here. This could be made more efficient by calculating which
+    # policies and resources we are missing, and selectively fetching, filtering, and updating the cache.
+    # This is not expected to cause an issue in production since in most cases, bulk new memberships would imply
+    # bulk new groups, which shouldn't have much if any policies associated to them.
+    previously_connectable = cache.connectable_resources
+
+    # Use the previous connectable IDs so that the recomputation yields the difference
+    cache = %{hydrate(client) | connectable_resources: previously_connectable}
+
+    recompute_connectable_resources(cache, client)
+  end
+
+  @doc """
+    Removes all policies, resources, and memberships associated with the given group_id from the cache.
+  """
+
+  @spec delete_membership(t(), Actors.Membership.t(), Clients.Client.t()) ::
+          {:ok, [Cache.Cacheable.Resource.t()], [Ecto.UUID.t()], t()}
+
+  def delete_membership(cache, membership, client) do
+    gid_bytes = dump!(membership.group_id)
+
+    rid_bytes_to_remove =
+      cache.policies
+      |> Enum.filter(fn {_id, p} -> p.actor_group_id == gid_bytes end)
+      |> Enum.map(fn {_id, p} -> p.resource_id end)
+      |> Enum.uniq()
+
+    updated_policies =
+      cache.policies
+      |> Enum.reject(fn {_id, p} -> p.actor_group_id == gid_bytes end)
+      |> Enum.into(%{})
+
+    updated_resources =
+      cache.resources
+      |> Map.drop(rid_bytes_to_remove)
+
+    updated_memberships =
+      cache.memberships
+      |> Map.delete(gid_bytes)
+
+    cache = %{
+      cache
+      | policies: updated_policies,
+        resources: updated_resources,
+        memberships: updated_memberships
+    }
+
+    recompute_connectable_resources(cache, client)
+  end
```

Comment:
There's a lot going on here, so double check my logic here...

Is there a bug here that might crash the process?

Here's my thought process:

- The resource gets removed from `cache.resources` (lines 243-245 of domain/cache/client.ex)
- Then `conforming_resource_ids` will find that resource ID because other policies still reference it (lines 547-551 of domain/cache/client.ex)
- `adapted_resources` tries to get that resource with `Map.get(resources, id)`, where `resources` comes from the cache, but returns `nil` because it got removed in step 1  (line 535 of domain/cache/client.ex)
- The `nil` gets rejected at line 537 of domain/cache/client.ex